### PR TITLE
wait for `connected` event from ApiPromise while reconnecting

### LINKED ISSUE #1812: When a connection is lost, it can never be reconnected
### Prerequisites

* [x] Are you running the latest version(s)? v2.6.1
* [ ] Did you check the [debugging guide](https://doc.subquery.network/academy/tutorials_examples/debug-projects.html#how-to-debug-a-subquery-project)?
       No, Can not open this page
      image
* [ ] Did you check the [FAQs](https://doc.subquery.network/faqs/faqs.html) and [Discussions](https://github.com/subquery/subql/discussions)?
      No, can not open this page
     image
* [x] Are you reporting to the correct repository?
* [x] Did you search for an existing issue or pull request?

### Description

[Description of the bug or feature]

When a connection is lost, it can never be reconnected

### Details

These details can help to reproduce the environment the issue is occurring

**Local Environment:** [You can get this information from executing `subql version`.]
      DB_USER: postgres
      DB_PASS: ${DB_PASSWORD}
      DB_DATABASE: subql_bridger_darwinia
      DB_HOST: postgres
      DB_PORT: 5432

**Query Version:** [What is the version of the query service?]
onfinality/subql-query:v2.0.0
**Indexer Version:** [What is the version of the indexer service?]
onfinality/subql-node:v2.6.1
**Network Details:**
  * [Network] 
     Darwinia/Crab/Kusama/Polkadot/Rococo/Moonbeam/Pangolin/Pangoro
  * [Block height, a block height where the issue is happening]
     Any block. Whenever the connection is lost
  * [Dictionary endpoint, if used]

### Steps to Reproduce

1. Run a node
2. Wait..

**Example project:** [A link to a minimal example that can reproduce the issue]

**Expected behavior:** [What you expected to happen]

**Actual behavior:** [What actually happened]

### Any other information

```

2023-06-20T12:11:52.372Z <connection-pool> INFO No suspended endpoints.
2023-06-20T12:12:02.038Z <BlockDispatcherService> INFO Enqueueing blocks 396981...396981, total 1 blocks
2023-06-20T12:12:02.038Z <BlockDispatcherService> INFO fetch block [396981,396981], total 1 blocks
2023-06-20T12:12:04.042Z <benchmark> INFO 0.08 blocks/s. Target height: 396,981. Current height: 396,981. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20T12:12:12.190Z <BlockDispatcherService> INFO Enqueueing blocks 396982...396982, total 1 blocks
2023-06-20T12:12:12.190Z <BlockDispatcherService> INFO fetch block [396982,396982], total 1 blocks
2023-06-20T12:12:19.042Z <benchmark> INFO 0.10 blocks/s. Target height: 396,982. Current height: 396,982. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20T12:12:27.597Z <BlockDispatcherService> INFO Enqueueing blocks 396983...396983, total 1 blocks
2023-06-20T12:12:27.598Z <BlockDispatcherService> INFO fetch block [396983,396983], total 1 blocks
2023-06-20T12:12:34.042Z <benchmark> INFO 0.07 blocks/s. Target height: 396,983. Current height: 396,983. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20T12:12:37.600Z <BlockDispatcherService> INFO Enqueueing blocks 396984...396984, total 1 blocks
2023-06-20T12:12:37.600Z <BlockDispatcherService> INFO fetch block [396984,396984], total 1 blocks
2023-06-20T12:12:49.043Z <benchmark> INFO 0.10 blocks/s. Target height: 396,984. Current height: 396,984. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20T12:12:52.374Z <connection-pool> INFO No suspended endpoints.
2023-06-20T12:12:52.606Z <BlockDispatcherService> INFO Enqueueing blocks 396985...396985, total 1 blocks
2023-06-20T12:12:52.607Z <BlockDispatcherService> INFO fetch block [396985,396985], total 1 blocks
2023-06-20T12:13:02.610Z <BlockDispatcherService> INFO Enqueueing blocks 396986...396986, total 1 blocks
2023-06-20T12:13:02.610Z <BlockDispatcherService> INFO fetch block [396986,396986], total 1 blocks
2023-06-20T12:13:04.043Z <benchmark> INFO 0.08 blocks/s. Target height: 396,986. Current height: 396,986. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20T12:13:12.611Z <BlockDispatcherService> INFO Enqueueing blocks 396987...396987, total 1 blocks
2023-06-20T12:13:12.611Z <BlockDispatcherService> INFO fetch block [396987,396987], total 1 blocks
2023-06-20T12:13:19.043Z <benchmark> INFO 0.10 blocks/s. Target height: 396,987. Current height: 396,987. Estimated time remaining: 0 days 00 hours 00 mins
2023-06-20 12:13:22          API-WS: disconnected from wss://rpc.darwinia.network: 1006:: Abnormal Closure
2023-06-20T12:13:22.112Z <connection-pool> WARN disconnected from wss://rpc.darwinia.network
2023-06-20T12:13:22.171Z <connection-pool> ERROR unable to reconnect to endpoint wss://rpc.darwinia.network
2023-06-20T12:13:26.060Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:26.081Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:31.059Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:31.080Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:34.043Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:13:36.059Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:36.080Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:41.060Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:41.080Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:46.060Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:46.080Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:49.042Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:13:51.061Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:51.081Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:52.373Z <connection-pool> INFO No suspended endpoints.
2023-06-20T12:13:56.060Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:13:56.081Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:01.060Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:01.082Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:04.042Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:14:06.061Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:06.081Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:11.061Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:11.082Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:16.063Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:16.082Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:19.043Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:14:21.008Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:21.062Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:21.083Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:21.388Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:26.063Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:26.083Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:31.040Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:31.062Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:31.085Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:31.399Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:34.043Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:14:36.062Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:36.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:41.058Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:41.062Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:41.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:41.423Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:46.062Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:46.085Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:49.043Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:14:51.070Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:51.076Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:51.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:51.434Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:14:52.373Z <connection-pool> INFO No suspended endpoints.
2023-06-20T12:14:56.071Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:14:56.085Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:01.071Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:01.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:01.109Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:15:01.445Z <health> ERROR undefined Error: Endpoint is not healthy
2023-06-20T12:15:04.043Z <benchmark> INFO Fully synced, waiting for new blocks
2023-06-20T12:15:06.071Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:06.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:11.071Z <FetchService> ERROR Having a problem when getting finalized block Error: All of the endpoints are suspended at the moment
2023-06-20T12:15:11.086Z <FetchService> ERROR Having a problem when getting best block Error: All of the endpoints are suspended at the moment
```

```
version: "3"

services:
  node-subql-bridger-kusama:
    container_name: node-subql-bridger-kusama
    image: onfinality/subql-node:v1.14.1
    restart: always
    environment:
      DB_USER: postgres
      DB_PASS: ${DB_PASSWORD}
      DB_DATABASE: subql_bridger_kusama
      DB_HOST: postgres
      DB_PORT: 5432
    command:
      - --disable-historical=true
      - -f=ipfs://QmRWWjGDCrtn9gC35zpHhiEK3vuc2JumD5hxtihSzUAwU2
      - --db-schema=app
      - --network-endpoint=wss://kusama-rpc.polkadot.io

    depends_on:
      - postgres
    healthcheck:
      test: ["CMD", "curl", "-f", "http://node-subql-bridger-kusama:3000/ready"]
      interval: 3s
      timeout: 5s
      retries: 10
    networks:
      - subql

```

### PR DESCRIPTION
**Pull request by DeveloperInProgress**
# Description
Wait for `connected` event from ApiPromise before resolving apiConnect()

Fixes #1812 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/node-core/src/indexer/connectionPool.service.ts [hunk 1]
```diff
@@ -185,19 +189,19 @@ export class ConnectionPoolService<T extends IApiConnectionSpecific<any, any, an
...
+    logger.warn(`disconnected from ${this.pool[apiIndex].endpoint}`);
 
     try {
-      logger.debug(`reconnecting to ${endpoint}...`);
+      logger.debug(`reconnecting to ${this.pool[apiIndex].endpoint}...`);
       await this.connectToApi(apiIndex);
     } catch (e) {
-      logger.error(`unable to reconnect to endpoint ${endpoint}`, e);
+      logger.error(`unable to reconnect to endpoint ${this.pool[apiIndex].endpoint}: ${e}`);
+      delete this.pool[apiIndex];
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.ts:200]: So we only retry connecting once? Why not try multiple times with exponential backoff.

File: packages/node-core/src/indexer/connectionPool.service.ts [hunk 2]
```diff
@@ -138,6 +138,10 @@ export class ConnectionPoolService<T extends IApiConnectionSpecific<any, any, an
   }
 
   private getNextConnectedApiIndex(): number | undefined {
+    if (Object.keys(this.pool).length === 0) {
+      throw new Error(`No connected endpoints available`);
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.ts:142]: This is ambiguous. I cant tell if it means that all the connected endpoints are available but suspended or we have no connections to any endpoint.

**Review by stwiname — COMMENTED** [review]
Its really hard to have any confidence with this without having tests. Can you please add tests for all the connection pool logic. 

I suggest mocking something rather than using ApiPromise.

File: packages/node-core/src/indexer/connectionPool.service.ts [hunk 3]
```diff
@@ -185,19 +193,42 @@ export class ConnectionPoolService<T extends IApiConnectionSpecific<any, any, an
...
+          );
+          delete this.pool[index];
+        }
+      }
+    };
 
-    logger.info(`reconnected to ${endpoint}!`);
+    (async () => {
+      await tryReconnect();
+    })();
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.ts:231]: Whats going on here?

File: packages/node-core/src/indexer/connectionPool.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        message: 'Connection error',
+      });
+      connectionPoolService.handleApiError(0, {
+        name: 'ConnectionError',
+        errorType: ApiErrorType.Connection,
+        message: 'Connection error',
+      });
+
+      // Wait for a while to see if handleApiDisconnects gets called multiple times
+      await new Promise((resolve) => setTimeout(resolve, 5000));
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.spec.ts:108]: We have a `delay` function

File: packages/node-core/src/indexer/connectionPool.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      await waitFor(() => (mockApiConnection.apiConnect as any).mock.calls.length === 5);
+      expect(mockApiConnection.apiConnect).toHaveBeenCalledTimes(5);
+      expect(connectionPoolService.numConnections).toBe(0);
+    }, 31000);
+  });
+
+  describe('handleApiDisconnects', () => {
+    // ...
+
+    it('should call handleApiDisconnects only once when multiple connection errors are triggered', async () => {
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.spec.ts:87]: i think this test is flawed. The mock connection will connect instantly https://github.com/subquery/subql/pull/1824/files#diff-8901f6b7ece4aeb1181e733a8b078308404b5638c239320b2bc53ad972233b12R24 but it needs to take some time like in the polkadot api

Comment (DeveloperInProgress) [reply packages/node-core/src/indexer/connectionPool.service.spec.ts:87]: https://github.com/subquery/subql/pull/1824/files#diff-8901f6b7ece4aeb1181e733a8b078308404b5638c239320b2bc53ad972233b12R89

I override it to connect after 10000ms

File: packages/node-core/src/indexer/connectionPool.service.ts [hunk 6]
```diff
@@ -185,19 +193,40 @@ export class ConnectionPoolService<T extends IApiConnectionSpecific<any, any, an
     return Object.keys(this.pool).length;
   }
 
-  async handleApiDisconnects(apiIndex: number, endpoint: string): Promise<void> {
-    logger.warn(`disconnected from ${endpoint}`);
-    delete this.pool[apiIndex];
+  async handleApiDisconnects(index: number) {
```
Comment (stwiname) [inline packages/node-core/src/indexer/connectionPool.service.ts:196]: I think it makes sense for this to be made private now. It only gets called internally and making it private will also ensure that we update the other SDKs properly

**Comment by stwiname** [comment]
Whats the easiest way I can test these improvements out?

**Comment by DeveloperInProgress** [comment]
> Whats the easiest way I can test these improvements out?

```
if(Math.random() < 0.01) {
    await api.disconnect();
  }
```

I added this to the start of `fetchBlockBatches` to trigger occasional disconnects

**Pull request merged**